Blog Payjoin Probing Attacks (#20)
payjoin/payjoindevkit.org: docs/_blog/payjoin-probing-attacks.md

---
title: "Payjoin Probing Attacks: Facts, Mitigations, and Why Payjoin Still Wins for Privacy"
description: ""
date: "2025-03-31"
authors:
  - Dan Gould
tags:
  - privacy
  - integrations
---
<h2></h2>

The following is a conclusive address of concerns around UTXO probing attacks on

> Payjoin, clarification of why current mitigations are effective, and difinitive

Reviewer: ```suggestion
Payjoin, clarification of why current mitigations are effective, and definitive
```

By specification, automated payment processors operating as Payjoin receivers must broadcast the
sender's valid fallback transaction after an expiration timeout if
receivers don't see their Payjoin broadcast. This enforces a
minimum relay fee rate policy by default.

The mechanism may also be implemented by manual peer-to-peer receivers
as an additional precaution. A broadcast may be triggered manually or
automatically after an appropriate amount of time passes, for example, after an Async
Payjoin session expires.

### Input Replay

Automated payment processors should respond to senders using fallback transactions containing
the same sender inputs with the same receiver input of their own. This secures
the floor cost imposed by the above mitigations.

### Session Policy Binding

In Async Payjoin, receive sessions are bound to the bitcoin URI used to request them.
Receivers may bind specific policies to specific URIs given to
individual senders depending on the perceived level of trust.

## Refuting Core Concerns

- **Claim**: Probing is costless<br/>
  **Reality**: Attackers must pay fees or risk paying the receiver the minimum
  value when their fallback is broadcast.

> - **Claim**: Current mitigations are innefective<br/>

Reviewer: ```suggestion
- **Claim**: Current mitigations are ineffective<br/>
```

- **Improves global privacy**: Payjoins can look like non-payjoins,
  compounding difficulty for surveillance adversaries.

## Edge Cases and Real-World Context

Attacks exist in theory, particularly if implementation guidelines aren't
followed. Such cases are implementation issues — not Payjoin protocol flaws. The
Payjoin Dev Kit guides implementators to make misuse difficult.

While subtle attacks like dusting do occur on Bitcoin, probing Payjoin at scale
is costly and practically limited.

## How to Secure your Payjoin Implementation

- Use Payjoin Dev Kit to follow best practices by default.
- For custom implementations, follow the BIP guidelines and mitigation checklist in this article.
- Ask the Payjoin Dev Kit team for help to ensure your implementation takes
  appropriate precautions like request persistence and fallback broadcast.

## Why Payjoin Wins

Payjoin's savings and privacy benefits vastly outweigh theoretical probing
risks. Adopting Payjoin offers good will to the entire bitcoin network, since
higher rate of payjoin usage increases privacy for even non-payjoin
transactions. Basic implementation policy and best practices offered by Payjoin

> Dev Kit mean thatno matter how big of a problem you think probing is, there's still a

Reviewer: ```suggestion
Dev Kit mean that no matter how big of a problem you think probing is, there's still a
```

> receiver implementation where apples-to-apples benefits exceed costs and it is rational to Payjoin.

Reviewer: This last sentence is clunky and imo should be reworked to end on a strong note/be less passive.

"PDK offers best implementation policy and best practices that address all known concerns around probing, and make it economically rational to receive with Payjoin."
